### PR TITLE
Isothermal, Polytropic and Multistage compressor

### PR DESCRIPTION
**Pull request by BenPortner**
- implements new class `IsothermalCompressor` which can calculate the theoretical compression power for the isothermal limit
- implements test for new `IsothermalCompressor` class
- implements common baseclass for `IsothermalCompressor` and `IsentropicCompressor`
- allows users to enforce a compressor type
- compressor type is now listed in `design_parameters`

### CODE REVIEW

**Review by yoelcortes — COMMENTED** [review]
Hi Ben,

Looking good! 

I am not too familiar with compressors, but would a heat exchanger be needed to cool down an isothermal compressor? I can take care of adding a heat exchanger since it is a little more complex to make it run as an "auxiliary unit operation" like in https://github.com/BioSTEAMDevelopmentGroup/biosteam/blob/30d3594923520a210cf82bd69957001a1cfed870/biosteam/units/_flash.py#L221

I plan on making a few other enhancements before merging, but I'll get to it a little later since I'm traveling until the 4th of August.

Thanks for contributing your efforts!

**Comment by BenPortner** [comment]
Hi @yoelcortes,

An isothermal compressor is a theoretical construct. Isothermal compression can be approached by compressing in multiple stages with intermittent cooling but true isothermal behavior is never reached (the isothermal limit is approached as the number of stages approaches infinity). For this reason, it does not really make sense to implement a HX in my opinion. It just wouldn't make it any more "realistic" 😉 

Cheers.
Ben

**Comment by yoelcortes** [comment]
Sounds good, thanks for the explanation!

**Comment by BenPortner** [comment]
Update @yoelcortes:
- merged master into this branch such that everything is up to date
- added a new unit `PolytropicCompressor` to simulate polytropic compression

**Comment by BenPortner** [comment]
Update @yoelcortes:

- added a new unit `MultistageCompressor` to simulate multistage compression
- added methods to support `+` operator and `sum` for `HeatUtility` and `PowerUtility`

**Comment by yoelcortes** [comment]
Hi Ben,

I am super happy with all the work you have contributed. It is thanks to you that we finally have some compressor tools. I am usually not picky about code style, but since this is one of the first major pull requests, I made changes here just to show my preferences, including:
* Only comments as needed (this is personal preference, but good to keep in mind). 
* Avoid duplicated design results for power and duty (due to how biosteam automatically adds them).
* Avoid outlet properties that are not explicitly used for design/costing algorithm (e.g. temperature) from being in design_results.

Knowing you already made unit tests, I am sorry I did not share with you my thoughts on these earlier. From here on I won't be too picky about what things are added to the dictionaries or comments. So please feel free to code freely.

For __add__ methods, I feel more strictly that they should return only objects of the same type, so I removed that potential behavior you added recently. Also, although not a major overhead, using python sum to add many heat utilities is much slower compared to using the `sum_by_agent` method once. It may not be as beautiful, but I think it is best in the long run.

Thanks!

**Comment by BenPortner** [comment]
Hi @yoelcortes,

thanks for the feedback and good job implementing the cost algorithm! I really enjoy contributing to biosteam and I have to say I am very impressed by your development speed and user support. This is certainly unique for a small reasearch-driven open source project like biosteam! Yet, I believe there are some issues that need to be addressed: 

# General comments:
1) Your commits introduce changes, which cause 11 tests to fail. I think our primary focus before introducing new features should be not to break existing features.
2) Your commit f44f03c is super comprehensive and the changes are not described well in the commit message. This makes it very hard to review (which might be necessary to fix the tests).

My suggestion to avoid 1) and 2): 
- Committing more frequently
- Each commit should touch only one part of the code
- Testing after each commit. If tests fail, they should be fixed before introducing further changes

3) It seems your commits somehow meddled with the git history. Commit https://github.com/BioSTEAMDevelopmentGroup/biosteam/pull/108/commits/283dbbd39c421072b6643c1558d6f92e3ef2a95e shows major parts of my code and attributes them to you. I don't know what went wrong there but I would like to see this commit reverted. Although biosteam is open source, I believe correct attribution of code pieces to their respective authors is only fair.

# Specific comments:

4) > Only comments as needed (this is personal preference, but good to keep in mind).

I will try to follow this recommendation. Could you give me an example of un-needed comments in my code to help me get a better grasp?


5) > Avoid duplicated design results for power and duty (due to how biosteam automatically adds them)

I disagree for three reasons:
- This is not true for the `IsothermalCompressor` class. This one has a duty, but no utility. Hence, biosteam will not add the duty automatically. 
- I believe it is always better to be explicit rather than implicit. For example, a novice user might not know that isentropic compressors have no duty by definition. He might look for the duty in the `design_results` and the `heat_utilities` and won't find it. He might think this is a bug. If the `design_results` show that the duty is zero, it is crystal clear that there is no duty by intention rather than accident.
- It is beneficial to have the same fields in `design_results` for all compressor classes because different compressors can be mixed in the `MultistageCompressor` class. Having different fields for different compressors will complicate the summary of design parameters across the stages.

6) > Avoid outlet properties that are not explicitly used for design/costing algorithm (e.g. temperature) from being in design_results.

Understood.

7) >Knowing you already made unit tests, I am sorry I did not share with you my thoughts on these earlier. From here on I won't be too picky about what things are added to the dictionaries or comments. So please feel free to code freely.

I don't understand how the dictionaries and comments relate to the unit tests. Could you clarify?

8) > For add methods, I feel more strictly that they should return only objects of the same type

I believe this was the case? Here is my code:

```python

# in PowerUtility:

    def __add__(self, other):
        return PowerUtility.sum([self,other]) # returns a PowerUtility

    def __radd__(self, other):
        return PowerUtility.sum([self,other]) # returns a PowerUtility

# in HeatUtility:

    def __add__(self, other):
        if hasattr(other, "__iter__"):
            return HeatUtility.sum_by_agent([self, *other]) # returns HeatUtility
        elif hasattr(other, "agent"):
            return HeatUtility.sum_by_agent([self, other]) # returns HeatUtility
        else:
            return self # returns HeatUtility

    def __radd__(self, other):
        return self + other # calls __add__, which will return a HeatUtility
```

Note that your solution does not work, i.e. break the test in `test_utilities.py`.

**Comment by yoelcortes** [comment]
@BenPortner, All the feedback and recommendations are much appreciated and well detailed. There are certainly things I hope to get better with as the lead developer. I hope to get back to it later today if not tomorrow.

If you'd like to keep working on a separate brach to avoid pulling now that's is OK and I can merge any commits from you branch into this one later (once I've addressed some issues you discussed).

Thank you!

**Comment by yoelcortes** [comment]
@BenPortner

### Regarding general comments:

All sounds good. I've been getting away with large commits because outside contribution has been stagnant for a while. I'll get back to the better habits you suggested. Thanks for your thoughts and recommendations.

Regarding your last comment, there was a merge conflict that makes the history hard to follow. I worked on the code while on a long flight and forgot to pull just before my flight. I do not understand why github would attribute such a large chunck of code to the user merging it. It is disappointing to know that github cannot discern contributions with merge conflicts. My goal is to keep high standards for authorship in BioSTEAM. Later next week, I will add to the "contributing" section of the documentation appropriate ways to avoid and handle merge conflicts. A note on having a merge commit then editing comment will be added too. Please note that the github history has not been edited.

By the way, I appreciate your company for investing resources and contributing to leverage BioSTEAM to their needs.

### Regarding specific comments:

1) With the reasonable variable and method names you use, I think a lot of your code do not need any comments to understand it. To some degree, some comments feel like documentation of other documented parts of biosteam. A good assumption I like to use is that the reader is familiar with common attributes/properties/methods used (and if not, they can check out the docstring). Here are a few examples where comments can be removed:

```python
# overwrite subcomponent id
ID = f"{self.ID}_{subcomponent.ticket_name}{i_stage}"
subcomponent.ID = ID
```

```python
# set heat utility
feed = self.ins[0]
out = self.outs[0]
u = bst.HeatUtility(heat_transfer_efficiency=1, heat_exchanger=None)
u(unit_duty=self.Q, T_in=feed.T, T_out=out.T)
```

```python
# save other design parameters
F_mol = self.outs[0].F_mol
self.design_results['Ideal Power'] = self.ideal_power ...
```

2) You are welcome to have the duty in the design dictionary if you feel it serves a good purpose. I still think it is better to not include duty or power in `design_results` dictionary. Following your points on this:
* In BioSTEAM, every unit with duty should have a heat utility to keep the energy balance (if a compressor class doesn't have one, we should go ahead and add it).
* There are benefits to only using heat utility and power utility objects to set and get values (instead of `design_results`) as they are the standard objects that manage utilities. The multistage compressor can access heat_utility and power utility objects from subcomponents instead. This actually takes place in the multistage compressor's `_run` method. Instead of going through every subcomponent design results to get the duty in `_design`, it is also possible to have `duty = sum([i.duty for i in self.heat_utilities])`.

If you do decide to keep the duty in here, don't forget about the heat transfer efficiency of HeatUtility objects :)

4) I noticed that you added a `test_compressor.py` module that tests items in `design_results` of compressors. We'll need to update these since I changed the `design_results` dictionaries (I was referring to `design_results` when I wrote dictionaries). Never mind the comments.

5) `HeatUtility.sum_by_agent` returns a tuple when the agent is not specified. In your tests, you'll find that the sum of heat utilities returns a tuple of heat utilities:
https://github.com/BioSTEAMDevelopmentGroup/biosteam/blob/3be4b0f92adf26105d3a1c5b05b8910c78bd90a9/tests/test_utilities.py#L23
Thanks for noting that the utility tests are broken and the changes I introduced were not working. I went ahead and fixed it along with the tests:
https://github.com/BioSTEAMDevelopmentGroup/biosteam/blob/94481962340c0c2c300d5a663ed4616491f702a6/tests/test_utilities.py#L23

As soon as we can get the tests passing, I would love to get this branch merged to master. Please feel free to work on this branch now (including adding stuff to design_results as you see fit). I can help with getting the compressor tests back and running, just let me know.

### Unrelated to any comments
1. To not break biosteam unit connections, you'll need to keep using the same inlets and outlets:

```python
# This would disconnect units and biosteam will think they are still connected.
# self._ins = self.compressors[0].ins
# self._outs = self.hxs[-1].outs

# Use something like this instead
self._ins[0].copy_like(self.compressors[0].ins)
for i, stream in enumerate(self.hxs[-1]): self._outs[i] = stream
```

2. To speed up recycle loop convergence, we can split the mass and energy balances from the design and costing. For example:
```python
def _run(self):
    ...
    # simulate all subcomponents
    for u in units:
        u.run() # 
...
def _design(self):
    ...
    for u in self.units: u._summary() # this does the design and costing
    self.power_utility.mix_from([u.power_utility for u in units])
    self.heat_utilities = bst.HeatUtility.sum_by_agent([h for u in units for h in u.heat_utilities])
```

Feel free to ignore point number 2 as it is related more to code optimization. I just thought it would be good to let your know

**Comment by yalinli2** [comment]
Late to the party but want to chime in a bit here - with regarding to the comment, I'll side with @BenPortner , I'd think being explicit is always better. While the comment might not be helpful to experienced users, they can be extremely helpful to new users. There will be new users looking at the codes/using them as the templates to learn (that's what I did a few years back and I'd appreciate the comments like these).

So I don't think we need to have comments like this for every module (I agree they might are very detailed), but I also don't think we need to delete them

**Comment by yoelcortes** [comment]
@yalinli2, right, certainly many trains of thought on this! Comments are definitely helpful. The main take away I'd like to share is that it is good to be mindful to avoid rewriting code. 

Thanks!

**Comment by BenPortner** [comment]
Hello @yoelcortes,

# General comments

Thank you for adding me to the authors list and for adding my name in the copyright notice. I was also a bit surprised how Github handled this case. However, with your changes and the git history I believe the contributions are clear enough. Thanks!

# Specific comments

## 1) In-code comments

I agree with @yalinli2 that more comments are usually better. For beginners, every comment documenting the intention of the code lines will be helpful. At the same time I don't see any disadvantages of thorough commenting. It shouldn't make the code any slower. And most modern IDEs allow collapsing continuous chunks of comment, so clarity shouldn't be an issue either. However, this is your project so I won't argue. I will try to reduce comments to a minimum from now on.

## 2) Duty in design_results

> In BioSTEAM, every unit with duty should have a heat utility to keep the energy balance (if a compressor class doesn't have one, we should go ahead and add it).

I still disagree with this point, see [my explanation of the `IsothermalCompressor` class](https://github.com/BioSTEAMDevelopmentGroup/biosteam/pull/108#issuecomment-1199687546).

More generally, the name `design_results` suggests that all relevant parameters that describe the final design of a unit are accessible from there. Duty, power and outlet temperature are fundamental design parameters of compressors. This means any user of the compressor classes will want to access them. Spreading these data across three (or more) different objects is plain user-unfriendly in my opinion. It makes the code longer and messier. It also makes it harder for beginners, because they might not know where to look for which datum. On the other hand, I don't see any disadvantage to storing these values in the design_results dict. Could you tell me *why* you don't want these in there?

``` python
# your suggestion
K.power_utility.rate
K.heat_utilities[0].duty + K.heat_utilities[1].duty + ...
K.outs[0].T

# my suggestion
K.design_parameters["Power"]
K.design_parameters["Duty"]
K.design_parameters["Outlet Temperature"]
```

On a related note: `HXprocess` does not document the duty *anywhere* in the results(). How should a user know that he/she has to call `HXprocess.Q` to get this information? Please, let's add the duty to the design_results!

## 3 compressor tests

I spent a whole day trying to fix the tests today. Several issues persists but I cannot put any more time into fixing. Can you please handle these:
- https://github.com/BioSTEAMDevelopmentGroup/thermosteam/issues/71
- `test_compressor_design` is supposed to test the cost correlations. Since I do not have access to these, can you re-design the function?
- `test_isothermal_hydrogen_compressor`: the assertion in line 132 fails because the flow rate of the coolant in the heat utility changed. This is easy to fix but before adjusting the number we should make sure that it is correct. It is strange that the flow rate changed because the duty is still the same and also the flow rate in all other tests remained the same. Could you check why it changed only here? 
- I adopted the `units.run` and `units.summary` approach for the `MultistageCompressor` as you suggested. However, this introduced a new problem: For some reason, the subcomponents are not assigned heat utilities during `summary`. This needs to be fixed before the docstrings can be corrected.

## 4 Heat utilities and their sums

I do *not* agree that `sum(list of hus)` should return an error. In general, I believe that the user should have the option to use expressions, even when there are alternative expressions, which are faster (in this case `bst.HeatUtility.sum_by_agent()`). In this particular example, I believe `sum` should work because a) this is what a beginner expects and b) this variant is more concise in my opinion.

More generally, I agree that __add__ should not return a tuple. Storing HUs in tuples (rather than a functional wrapper class) is a poor design choice. But this is the way biosteam works right now. All units return a tuple upon calling `unit.heat_utilities`. It is therefore only natural that `__add__` returns a tuple when HUs of different agents are summed. You accept that `bst.HeatUtility.sum_by_agent()` returns a tuple. Why not `__add__`?

## 5 breaking connections

> To not break biosteam unit connections, you'll need to keep using the same inlets and outlets:

```python
# This would disconnect units and biosteam will think they are still connected.
# self._ins = self.compressors[0].ins
# self._outs = self.hxs[-1].outs

# Use something like this instead
self._ins[0].copy_like(self.compressors[0].ins)
for i, stream in enumerate(self.hxs[-1]): self._outs[i] = stream
```

I deliberately wrote the code this way. Note that `self._ins = self.compressors[0].ins` does *not* reconnect `self.compressors[0].ins` (contrary to `self.ins = self.compressors[0].ins`). It merely makes sure that `self.ins[0]` and `self.compressors[0].ins[0]` point to the same `Stream` object (as they should because the multistage compressor inlet is the same as the inlet of the first stage!). Using `copy_like` is *not* the same, as there would still be two distinct streams, which can be manipulated independently (which shouldn't be possible because they are the same stream).

**Comment by yoelcortes** [comment]
@BenPortner,

Here are replies to your comments/questions:

1. Great!

2. Regarding duties:
* If you'd like to get the duty of unit operation more easy, I would suggest a property that accesses the heat utilities / power utility instead. For example: '<Unit>.net_duty' and '<Unit>.net_power and document this for beginners. We can also add these fields in the `<Unit>.results()` method for beginners. By minimizing data from being repeated throughout the code, I hope we can have less human errors when coding (e.g., not accounting for heat transfer efficiency) and save ram on the long run. 
* I understand that an isothermal compressor is a theoretical construct, but it should still close the energy balance for completion and comparison with realistic unit operations.
* HXprocess has no duty because it is does not use any heat utilities, it is used for heat integration and heat exchanger networks. Please use HXutility for utilities.

3. The bug in thermosteam has been fixed (thanks for the report!). I will look into your other points later tonight.

4. If we can get the associative law to work, I'd be OK with allowing heat utility objects and tuples to sum:
```python
# The associative rule has to be true regardless of heating agent
# Assume hus is a tuple of 5 heat utilities with 2 agents as follows
# * hus[0] and hus[3] is cooling water
# * The rest are low pressure steam
assert (hus[0] + hus[1]) + (hus[2] + hus[3] + hus[4]) == sum(hus)
# But (hus[0] + hus[1]) and (hus[2] + hus[3] + hus[4]) are both tuples
```
To get this to work we would need a but for this we would a new `HeatUtilities` object that is works as a tuple of heat utilities with __add__ and __radd__.

5. Let's say the inlet of a multistage compressor is connected to a flash vessel, then `self._ins = self.compressors[0].ins` would disconnect it. 

Let me know if you need any further clarification on these,
Thanks!

**Comment by yoelcortes** [comment]
Update: Unit objects now have `net_power` and `net_duty` properties :)

**Comment by yoelcortes** [comment]
All tests in `test_compressor.py` are now passing:
- `test_isothermal_hydrogen_compressor` was not passing because the duty increased due to the eta decreasing. I just divided by the eta to get the new flow.
- Multistage compressor is working now; just needed to set heat utilities after running `_summary` (not before).

I'll leave the docstrings to you.

Thanks!

**Review by yoelcortes — APPROVED** [review]
All looks good! There may still be enhancements/optimizations that can be done (and are welcome), but I believe this is ready to merge with the master branch. 

Thanks for the contribution!
